### PR TITLE
new-parametrized SWAP gate, remove Swap type

### PR DESCRIPTION
**Pull request by GiggleLiu**
Making `Swap` an alias of `PutBlock` as it should be

```julia
const Swap{N} = PutBlock{N,2,G} where G<:ConstGate.SWAPGate
const PSwap{N, T} = PutBlock{N,2,RotationGate{2,T,G}} where G<:ConstGate.SWAPGate
```

Now the `PSwap` is introduced, it is much faster that before.
Its interface is
```julia
    pswap(n::Int, i::Int, j::Int, α::Real)
    pswap(i::Int, j::Int, α::Real) -> f(n)
```

Note:
It depends on PR https://github.com/QuantumBFS/YaoArrayRegister.jl/pull/22

### CODE REVIEW

File: src/composite/put_block.jl [hunk 1]
```diff
@@ -104,3 +104,60 @@ function YaoBase.iscommute(x::PutBlock{N}, y::PutBlock{N}) where N
         return iscommute_fallback(x, y)
     end
 end
+
+const Swap{N} = PutBlock{N,2,G} where G<:ConstGate.SWAPGate
```
Comment (Roger-luo) [inline src/composite/put_block.jl:108]: can you change the printings as well?

**Comment by Roger-luo** [comment]
can you also add a benchmark for SWAP in `benchmarks`? so we could compare the performance before and after

File: src/composite/put_block.jl (hunk 1 above)
Comment (GiggleLiu) [reply src/composite/put_block.jl:108]: The printing is ok, you mean the printing for PSwap?

**Comment by GiggleLiu** [comment]
# the old swap gate
```julia
julia> swp = swap(16, 2, 7)
swap(2, 7)

julia> @benchmark reg |> swp
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     82.934 μs (0.00% GC)
  median time:      85.292 μs (0.00% GC)
  mean time:        91.830 μs (0.00% GC)
  maximum time:     324.066 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

# the old pswap gate
```julia
julia> @benchmark reg |> ps
BenchmarkTools.Trial: 
  memory estimate:  9.89 KiB
  allocs estimate:  133
  --------------
  minimum time:     573.222 μs (0.00% GC)
  median time:      623.330 μs (0.00% GC)
  mean time:        733.766 μs (1.73% GC)
  maximum time:     62.299 ms (98.93% GC)
  --------------
  samples:          6750
  evals/sample:     1
```

# the old swap gate (batchsize = 1000)
```julia
julia> @benchmark reg |> swp
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     3.169 ms (0.00% GC)
  median time:      4.886 ms (0.00% GC)
  mean time:        4.847 ms (0.00% GC)
  maximum time:     6.198 ms (0.00% GC)
  --------------
  samples:          1027
  evals/sample:     1
```

# the old pswap gate (batchsize = 1000)
```julia
julia> ps = put(9, (2,7)=>rot(SWAP, 0.5))
nqubits: 9
put on (2, 7)
└─ rot(SWAP gate, 0.5)

julia> @benchmark reg |> ps
BenchmarkTools.Trial: 
  memory estimate:  5.93 MiB
  allocs estimate:  128128
  --------------
  minimum time:     12.310 ms (0.00% GC)
  median time:      13.538 ms (0.00% GC)
  mean time:        14.614 ms (5.25% GC)
  maximum time:     82.989 ms (83.77% GC)
  --------------
  samples:          342
  evals/sample:     1

```

**Comment by GiggleLiu** [comment]
# the new swap gate
```julia
julia> ps = pswap(16, 2, 7, 0.5)
nqubits: 16
put on (2, 7)
└─ rot(SWAP gate, 0.5)

julia> @benchmark reg |> ps
BenchmarkTools.Trial: 
  memory estimate:  80 bytes
  allocs estimate:  2
  --------------
  minimum time:     130.900 μs (0.00% GC)
  median time:      133.472 μs (0.00% GC)
  mean time:        149.325 μs (0.00% GC)
  maximum time:     522.646 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

the new swap gate
julia> swp = swap(16, 2, 7)
nqubits: 16
swap(2, 7)
└─ SWAP gate

```julia
julia> @benchmark reg |> swp
BenchmarkTools.Trial: 
  memory estimate:  80 bytes
  allocs estimate:  2
  --------------
  minimum time:     80.877 μs (0.00% GC)
  median time:      81.943 μs (0.00% GC)
  mean time:        85.171 μs (0.00% GC)
  maximum time:     281.218 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```
# the new pswap gate (batch = 1000)
```julia
julia> ps = pswap(9, 2, 7, 0.5)
nqubits: 9
put on (2, 7)
└─ rot(SWAP gate, 0.5)

julia> @benchmark reg |> ps
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     6.979 ms (0.00% GC)
  median time:      8.058 ms (0.00% GC)
  mean time:        8.217 ms (0.00% GC)
  maximum time:     11.781 ms (0.00% GC)
  --------------
  samples:          607
  evals/sample:     1
```

# the new swap gate (batch=1000)
```julia
julia> swp = swap(9, 2, 7)
nqubits: 9
swap(2, 7)
└─ SWAP gate

julia> @benchmark reg |> swp
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     3.466 ms (0.00% GC)
  median time:      5.173 ms (0.00% GC)
  mean time:        5.113 ms (0.00% GC)
  maximum time:     6.181 ms (0.00% GC)
  --------------
  samples:          974
  evals/sample:     1
```

**Comment by GiggleLiu** [comment]
I will fix the printing. But it is so strange that current swap seems a bit slower than before, maybe it is just statistic error?

**Comment by Roger-luo** [comment]
I don't think so. It's over 5%, should be regression. And can you add this benchmark to benchmark test? I didn't write benchmark for swap before.

**Comment by GiggleLiu** [comment]
I would suggest removing printing specialization for swap, since we don't have them for other gates like `X` gate.

**Comment by GiggleLiu** [comment]
> I don't think so. It's over 5%, should be regression. And can you add this benchmark to benchmark test? I didn't write benchmark for swap before.

Any link to the benchmark test?

**Comment by Roger-luo** [comment]
The benchmarks are in benchmarks folder you can add swap as other gates. And we need to make swap fast again to merge this PR. It should be a regression since it is over 5%.

https://github.com/QuantumBFS/YaoBlocks.jl/tree/master/benchmark

**Comment by GiggleLiu** [comment]
```julia
julia> @benchmark $reg |> $swp
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     3.429 ms (0.00% GC)
  median time:      5.918 ms (0.00% GC)
  mean time:        5.484 ms (0.00% GC)
  maximum time:     7.430 ms (0.00% GC)
  --------------
  samples:          909
  evals/sample:     1
```

The number of samples is only 900, so the fluctuation is large, this is the old version again. I did't touch the implementation, so they sould have same performance.

**Comment by Roger-luo** [comment]
I could investigate it later, but can you put it into the benchmark suite first? so I could compare performance automatically.

**Comment by GiggleLiu** [comment]
I have added the benchmark, not sure it is ocrrect. Out of memory on my laptop.

**Comment by Roger-luo** [comment]
@GiggleLiu it seems `instruct!` is not correctly dispatch according to the CI report?

**Comment by GiggleLiu** [comment]
Should trigger the build again. It relies on lastest YaoArrayRegister patch.

**Comment by GiggleLiu** [comment]
Now I triggered the building, please merge it if possible?

**Comment by Roger-luo** [comment]
@GiggleLiu It seems the test cov falls a lot by this PR. can we have some test for that?

**Comment by GiggleLiu** [comment]
Ah, the newly added block is well tested though, must be some bug?

**Comment by Roger-luo** [comment]
Maybe, the new cov of other PR also falls

**Pull request merged**